File: lib/features/imports/csv/csv_import_metric.dart
/// What a CSV column can be mapped onto, and how its raw text becomes the
/// canonical (metric) value Health Connect stores.
///
/// v1 covers body composition only — the metrics a smart scale exports. Every
/// one is a single instant-in-time measurement, so a mapping needs exactly one
/// timestamp column and one column per metric. Interval records (steps, sleep,
/// workouts) need a second timestamp and are deliberately absent; adding them
/// later is an entry in [kCsvMetricCatalog] plus an end-timestamp column role,
/// not a rewrite.
library;

import '../../../core/presentation/measurement_input.dart';
import '../../../domain/health/health_permissions.dart';

/// A metric a CSV column can be mapped onto.
///
/// Every one is a single measurement at a single instant. Blood pressure is
/// absent on purpose: systolic and diastolic have to become ONE record, which
/// needs a two-columns-to-one-record rule the mapping model does not have.
enum CsvImportMetric {
  weight,
  bodyFat,
  leanBodyMass,
  boneMass,
  bodyWaterMass,
  height,
  basalMetabolicRate,
  heartRate,
  restingHeartRate,
  heartRateVariability,
  oxygenSaturation,
  respiratoryRate,
  bodyTemperature,
  basalBodyTemperature,
  bloodGlucose,
  vo2Max,
}

/// A unit a CSV column's numbers can be written in.
///
/// This is the unit of the *file*, chosen by the user per column. It is not the
/// app's display unit system: a file can be in pounds while the app displays
/// kilograms, and neither implies the other. Nothing here may consult
/// `unitSystemProvider` — see AGENTS.md invariant #2.
enum CsvUnit {
  kilograms,
  pounds,
  stones,
  grams,
  percent,
  fraction,
  centimeters,
  meters,
  inches,
  feet,
  kilocaloriesPerDay,
  kilojoulesPerDay,
  celsius,
  fahrenheit,
  beatsPerMinute,
  milliseconds,
  seconds,
  breathsPerMinute,
  millimolesPerLiter,
  milligramsPerDeciliter,
  millilitersPerKgPerMinute,
}

const double _kKilogramsPerStone = 6.35029318;
const double _kCentimetersPerFoot = 30.48;
const double _kKilojoulesPerKilocalorie = 4.184;

/// Health Connect stores blood glucose in mmol/L; the US convention is mg/dL.
/// The factor is the one `BloodGlucoseImportRecord.milligramsPerDeciliter`
/// already uses, kept here as its inverse so both directions agree.
const double _kMilligramsPerDeciliterPerMillimolePerLiter = 18.0;

/// How a column's raw number becomes the metric's canonical value.
sealed class CsvValueInterpretation {
  const CsvValueInterpretation();

  /// Whether resolving this needs the row's body weight as well as its own cell.
  bool get needsRowWeight => this is CsvMassShareOfWeight;
}

/// The number in the cell IS the metric, expressed in [unit].
final class CsvDirectValue extends CsvValueInterpretation {
  const CsvDirectValue(this.unit);

  final CsvUnit unit;

  @override
  bool operator ==(Object other) =>
      other is CsvDirectValue && other.unit == unit;

  @override
  int get hashCode => Object.hash(CsvDirectValue, unit);
}

/// The cell holds a **mass**, and the metric is that mass as a percentage of
/// the row's body weight.
///
/// Scales export "Fat mass (kg)" while Health Connect's `BodyFatRecord` stores a
/// percentage, so the percentage has to be derived — and it can only be derived
/// from the weight measured at the same moment, i.e. the weight column of the
/// SAME row. A row missing that weight cannot produce this metric, which is why
/// [CsvValueInterpretation.needsRowWeight] exists: the mapping validator asks
/// the interpretation, not the metric, whether a weight column is required.
final class CsvMassShareOfWeight extends CsvValueInterpretation {
  const CsvMassShareOfWeight(this.unit);

  /// Always a mass unit — the catalog only offers mass units for this.
  final CsvUnit unit;

  @override
  bool operator ==(Object other) =>
      other is CsvMassShareOfWeight && other.unit == unit;

  @override
  int get hashCode => Object.hash(CsvMassShareOfWeight, unit);
}

/// Everything the importer needs to know about one metric.
class CsvMetricSpec {
  const CsvMetricSpec({
    required this.targetType,
    required this.writePermission,
    required this.interpretations,
    required this.plausibleMin,
    required this.plausibleMax,
  });

  /// The Health Connect record class this produces, matching
  /// `ImportRecord.targetType`.
  final String targetType;

  /// The single Health Connect write permission this metric needs.
  final String writePermission;

  /// Offered in the UI in this order; the first is the default.
  final List<CsvValueInterpretation> interpretations;

  /// Bounds on the CANONICAL value, used to reject a row rather than write a
  /// number Health Connect would happily store. A 900 kg weight is a mis-mapped
  /// column, not a person.
  final double plausibleMin;
  final double plausibleMax;

  CsvValueInterpretation get defaultInterpretation => interpretations.first;
}

/// The metric catalog. Adding a metric is an entry here plus an ARB label.
final Map<CsvImportMetric, CsvMetricSpec> kCsvMetricCatalog = {
  CsvImportMetric.weight: CsvMetricSpec(
    targetType: 'WeightRecord',
    writePermission: HcPermissions.writeWeight,
    interpretations: const [
      CsvDirectValue(CsvUnit.kilograms),
      CsvDirectValue(CsvUnit.pounds),
      CsvDirectValue(CsvUnit.stones),
      CsvDirectValue(CsvUnit.grams),
    ],
    plausibleMin: 2,
    plausibleMax: 650,
  ),
  CsvImportMetric.bodyFat: CsvMetricSpec(
    targetType: 'BodyFatRecord',
    writePermission: HcPermissions.writeBodyFat,
    // The mass interpretations are what make a Withings-style export usable
    // without editing the file.
    interpretations: const [
      CsvDirectValue(CsvUnit.percent),
      CsvDirectValue(CsvUnit.fraction),
      CsvMassShareOfWeight(CsvUnit.kilograms),
      CsvMassShareOfWeight(CsvUnit.pounds),
      CsvMassShareOfWeight(CsvUnit.grams),
    ],
    // Below ~2% is not survivable and above ~75% is not anatomically possible;
    // both mean the weight column used for the derivation was the wrong one.
    plausibleMin: 2,
    plausibleMax: 75,
  ),
  CsvImportMetric.leanBodyMass: CsvMetricSpec(
    targetType: 'LeanBodyMassRecord',
    writePermission: HcPermissions.writeLeanBodyMass,
    interpretations: const [
      CsvDirectValue(CsvUnit.kilograms),
      CsvDirectValue(CsvUnit.pounds),
      CsvDirectValue(CsvUnit.grams),
    ],
    plausibleMin: 1,
    plausibleMax: 400,
  ),
  CsvImportMetric.boneMass: CsvMetricSpec(
    targetType: 'BoneMassRecord',
    writePermission: HcPermissions.writeBoneMass,
    interpretations: const [
      CsvDirectValue(CsvUnit.kilograms),
      CsvDirectValue(CsvUnit.pounds),
      CsvDirectValue(CsvUnit.grams),
    ],
    plausibleMin: 0.1,
    plausibleMax: 20,
  ),
  CsvImportMetric.bodyWaterMass: CsvMetricSpec(
    targetType: 'BodyWaterMassRecord',
    writePermission: HcPermissions.writeBodyWaterMass,
    interpretations: const [
      CsvDirectValue(CsvUnit.kilograms),
      CsvDirectValue(CsvUnit.pounds),
      CsvDirectValue(CsvUnit.grams),
    ],
    plausibleMin: 0.5,
    plausibleMax: 400,
  ),
  CsvImportMetric.height: CsvMetricSpec(
    targetType: 'HeightRecord',
    writePermission: HcPermissions.writeHeight,
    interpretations: const [
      CsvDirectValue(CsvUnit.centimeters),
      CsvDirectValue(CsvUnit.meters),
      CsvDirectValue(CsvUnit.inches),
      CsvDirectValue(CsvUnit.feet),
    ],
    // Canonical height is METRES (HeightImportRecord.meters).
    plausibleMin: 0.3,
    plausibleMax: 2.8,
  ),
  CsvImportMetric.basalMetabolicRate: CsvMetricSpec(
    targetType: 'BasalMetabolicRateRecord',
    writePermission: HcPermissions.writeBasalMetabolicRate,
    interpretations: const [
      CsvDirectValue(CsvUnit.kilocaloriesPerDay),
      CsvDirectValue(CsvUnit.kilojoulesPerDay),
    ],
    plausibleMin: 200,
    plausibleMax: 12000,
  ),

  // ── Vitals ────────────────────────────────────────────────────────────────
  // Each is one number at one instant, which is the only shape the mapping
  // model expresses. Ranges are "survivable human", not "clinically normal":
  // rejecting a real fever or a real bradycardia would be worse than storing it.
  CsvImportMetric.heartRate: CsvMetricSpec(
    targetType: 'HeartRateRecord',
    writePermission: HcPermissions.writeHeartRate,
    interpretations: const [CsvDirectValue(CsvUnit.beatsPerMinute)],
    plausibleMin: 1,
    plausibleMax: 300,
  ),
  CsvImportMetric.restingHeartRate: CsvMetricSpec(
    targetType: 'RestingHeartRateRecord',
    writePermission: HcPermissions.writeRestingHeartRate,
    interpretations: const [CsvDirectValue(CsvUnit.beatsPerMinute)],
    plausibleMin: 1,
    plausibleMax: 300,
  ),
  CsvImportMetric.heartRateVariability: CsvMetricSpec(
    targetType: 'HeartRateVariabilityRmssdRecord',
    writePermission: HcPermissions.writeHrv,
    interpretations: const [
      CsvDirectValue(CsvUnit.milliseconds),
      CsvDirectValue(CsvUnit.seconds),
    ],
    // Health Connect rejects RMSSD outside 1..200 ms.
    plausibleMin: 1,
    plausibleMax: 200,
  ),
  CsvImportMetric.oxygenSaturation: CsvMetricSpec(
    targetType: 'OxygenSaturationRecord',
    writePermission: HcPermissions.writeSpO2,
    interpretations: const [
      CsvDirectValue(CsvUnit.percent),
      CsvDirectValue(CsvUnit.fraction),
    ],
    plausibleMin: 50,
    plausibleMax: 100,
  ),
  CsvImportMetric.respiratoryRate: CsvMetricSpec(
    targetType: 'RespiratoryRateRecord',
    writePermission: HcPermissions.writeRespiratoryRate,
    interpretations: const [CsvDirectValue(CsvUnit.breathsPerMinute)],
    plausibleMin: 1,
    plausibleMax: 100,
  ),
  CsvImportMetric.bodyTemperature: CsvMetricSpec(
    targetType: 'BodyTemperatureRecord',
    writePermission: HcPermissions.writeBodyTemperature,
    interpretations: const [
      CsvDirectValue(CsvUnit.celsius),
      CsvDirectValue(CsvUnit.fahrenheit),
    ],
    plausibleMin: 25,
    plausibleMax: 45,
  ),
  CsvImportMetric.basalBodyTemperature: CsvMetricSpec(
    targetType: 'BasalBodyTemperatureRecord',
    writePermission: HcPermissions.writeBasalBodyTemperature,
    interpretations: const [
      CsvDirectValue(CsvUnit.celsius),
      CsvDirectValue(CsvUnit.fahrenheit),
    ],
    plausibleMin: 25,
    plausibleMax: 45,
  ),
  CsvImportMetric.bloodGlucose: CsvMetricSpec(
    targetType: 'BloodGlucoseRecord',
    writePermission: HcPermissions.writeBloodGlucose,
    interpretations: const [
      CsvDirectValue(CsvUnit.millimolesPerLiter),
      CsvDirectValue(CsvUnit.milligramsPerDeciliter),
    ],
    // Canonical is mmol/L: roughly 18..900 mg/dL.
    plausibleMin: 1,
    plausibleMax: 50,
  ),
  CsvImportMetric.vo2Max: CsvMetricSpec(
    targetType: 'Vo2MaxRecord',
    writePermission: HcPermissions.writeVo2Max,
    interpretations: const [
      CsvDirectValue(CsvUnit.millilitersPerKgPerMinute),
    ],
    plausibleMin: 5,
    plausibleMax: 100,
  ),
};

/// Converts [value] from [unit] to the metric's canonical unit — kg for masses,
/// metres for height, kcal/day for BMR, percent for body fat.
///
/// Conversion factors come from [MeasurementInput]'s shared constants; a local
/// constant here would be the "bare imperial check in a feature file" AGENTS.md
/// invariant #2 bans. Stones, feet and kJ have no constant there because no
/// entry screen offers them, so they are defined once above.
double convertCsvValueToCanonical(double value, CsvUnit unit) => switch (unit) {
      CsvUnit.kilograms => value,
      CsvUnit.pounds => value / kPoundsPerKilogram,
      CsvUnit.stones => value * _kKilogramsPerStone,
      CsvUnit.grams => value / 1000,
      CsvUnit.percent => value,
      CsvUnit.fraction => value * 100,
      CsvUnit.centimeters => value / 100,
      CsvUnit.meters => value,
      CsvUnit.inches => value * kCentimetersPerInch / 100,
      CsvUnit.feet => value * _kCentimetersPerFoot / 100,
      CsvUnit.kilocaloriesPerDay => value,
      CsvUnit.kilojoulesPerDay => value / _kKilojoulesPerKilocalorie,
      CsvUnit.celsius => value,
      CsvUnit.fahrenheit =>
        (value - kFahrenheitFreezingPoint) / kFahrenheitPerCelsius,
      CsvUnit.beatsPerMinute => value,
      CsvUnit.milliseconds => value,
      CsvUnit.seconds => value * 1000,
      CsvUnit.breathsPerMinute => value,
      CsvUnit.millimolesPerLiter => value,
      CsvUnit.milligramsPerDeciliter =>
        value / _kMilligramsPerDeciliterPerMillimolePerLiter,
      CsvUnit.millilitersPerKgPerMinute => value,
    };

/// Unit tokens recognised in a column header, longest first so `kcal` is not
/// matched by `cal` and `lbs` is not matched by `lb`.
const List<(String, CsvUnit)> _kHeaderUnitTokens = [
  ('kilograms', CsvUnit.kilograms),
  ('kilogram', CsvUnit.kilograms),
  ('pounds', CsvUnit.pounds),
  ('ml/kg/min', CsvUnit.millilitersPerKgPerMinute),
  ('mmol/l', CsvUnit.millimolesPerLiter),
  ('mg/dl', CsvUnit.milligramsPerDeciliter),
  ('breaths/min', CsvUnit.breathsPerMinute),
  ('brpm', CsvUnit.breathsPerMinute),
  ('kcal', CsvUnit.kilocaloriesPerDay),
  ('kj', CsvUnit.kilojoulesPerDay),
  ('bpm', CsvUnit.beatsPerMinute),
  ('lbs', CsvUnit.pounds),
  ('lb', CsvUnit.pounds),
  ('st', CsvUnit.stones),
  ('kg', CsvUnit.kilograms),
  ('cm', CsvUnit.centimeters),
  ('ms', CsvUnit.milliseconds),
  ('°c', CsvUnit.celsius),
  ('°f', CsvUnit.fahrenheit),
  ('in', CsvUnit.inches),
  ('ft', CsvUnit.feet),
  ('%', CsvUnit.percent),
  ('c', CsvUnit.celsius),
  ('f', CsvUnit.fahrenheit),
  ('g', CsvUnit.grams),
  ('m', CsvUnit.meters),
  ('s', CsvUnit.seconds),
];

/// The unit named in [header], or null when it names none.
///
/// Reads a unit off a label the user has ALREADY chosen to map — it never maps a
/// header string to a metric, so this is a default, not a vendor preset. Only
/// the parenthesised tail is considered, so a column called "Weight in grams of
/// food" cannot be read as grams.
CsvUnit? detectCsvUnitInHeader(String header) {
  final match = RegExp(r'\(([^()]*)\)\s*$').firstMatch(header.trim());
  if (match == null) return null;
  final inner = match.group(1)!.trim().toLowerCase();
  if (inner.isEmpty) return null;
  for (final (token, unit) in _kHeaderUnitTokens) {
    if (inner == token) return unit;
  }
  return null;
}

/// [spec]'s offered interpretation whose unit is [unit], preferring a direct
/// reading over a derived one. Null when the metric cannot express that unit.
CsvValueInterpretation? interpretationForUnit(CsvMetricSpec spec, CsvUnit unit) {
  for (final interpretation in spec.interpretations) {
    if (interpretation is CsvDirectValue && interpretation.unit == unit) {
      return interpretation;
    }
  }
  for (final interpretation in spec.interpretations) {
    if (interpretation is CsvMassShareOfWeight && interpretation.unit == unit) {
      return interpretation;
    }
  }
  return null;
}

File: app/src/main/kotlin/tech/mmarca/openvitals/features/imports/csv/CsvImportMetric.kt
package tech.mmarca.openvitals.features.imports.csv

import androidx.health.connect.client.permission.HealthPermission
import androidx.health.connect.client.records.BasalBodyTemperatureRecord
import androidx.health.connect.client.records.BasalMetabolicRateRecord
import androidx.health.connect.client.records.BloodGlucoseRecord
import androidx.health.connect.client.records.BodyFatRecord
import androidx.health.connect.client.records.BodyTemperatureRecord
import androidx.health.connect.client.records.BodyWaterMassRecord
import androidx.health.connect.client.records.BoneMassRecord
import androidx.health.connect.client.records.HeartRateRecord
import androidx.health.connect.client.records.HeartRateVariabilityRmssdRecord
import androidx.health.connect.client.records.HeightRecord
import androidx.health.connect.client.records.LeanBodyMassRecord
import androidx.health.connect.client.records.OxygenSaturationRecord
import androidx.health.connect.client.records.Record
import androidx.health.connect.client.records.RespiratoryRateRecord
import androidx.health.connect.client.records.RestingHeartRateRecord
import androidx.health.connect.client.records.StepsRecord
import androidx.health.connect.client.records.Vo2MaxRecord
import androidx.health.connect.client.records.WeightRecord
import kotlin.reflect.KClass

/**
 * What a CSV column can be mapped onto, and how its text becomes the
 * canonical value. Every metric is one number at one instant, except
 * [STEPS], the first interval record: an END_TIMESTAMP column ends each
 * row's span, and a missing end means one minute. Blood pressure is absent:
 * it needs two columns per record.
 *
 * The order, catalog and conversions mirror the Flutter build, so both apps
 * produce the same clientRecordIds. STEPS keys on the interval's start.
 */
enum class CsvImportMetric {
    WEIGHT,
    BODY_FAT,
    LEAN_BODY_MASS,
    BONE_MASS,
    BODY_WATER_MASS,
    HEIGHT,
    BASAL_METABOLIC_RATE,
    HEART_RATE,
    RESTING_HEART_RATE,
    HEART_RATE_VARIABILITY,
    OXYGEN_SATURATION,
    RESPIRATORY_RATE,
    BODY_TEMPERATURE,
    BASAL_BODY_TEMPERATURE,
    BLOOD_GLUCOSE,
    VO2_MAX,
    STEPS,
}

/** A unit a column's numbers are written in. The file's unit, not the app's display unit. */
enum class CsvUnit {
    KILOGRAMS,
    POUNDS,
    STONES,
    GRAMS,
    PERCENT,
    FRACTION,
    CENTIMETERS,
    METERS,
    INCHES,
    FEET,
    KILOCALORIES_PER_DAY,
    KILOJOULES_PER_DAY,
    CELSIUS,
    FAHRENHEIT,
    BEATS_PER_MINUTE,
    MILLISECONDS,
    SECONDS,
    BREATHS_PER_MINUTE,
    MILLIMOLES_PER_LITER,
    MILLIGRAMS_PER_DECILITER,
    MILLILITERS_PER_KG_PER_MINUTE,
    COUNT,
}

// The same constants the manual-entry screens use.
private const val PoundsPerKilogram = 2.2046226218
private const val CentimetersPerInch = 2.54
private const val KilogramsPerStone = 6.35029318
private const val CentimetersPerFoot = 30.48
private const val KilojoulesPerKilocalorie = 4.184
private const val FahrenheitFreezingPoint = 32.0
private const val FahrenheitPerCelsius = 1.8

/** Health Connect stores glucose in mmol/L; the US convention is mg/dL. */
private const val MilligramsPerDeciliterPerMillimolePerLiter = 18.0

/** How a column's raw number becomes the metric's canonical value. */
sealed interface CsvValueInterpretation {
    /** Whether resolving this needs the row's body weight as well as its own cell. */
    val needsRowWeight: Boolean get() = this is CsvMassShareOfWeight
}

/** The number in the cell IS the metric, expressed in [unit]. */
data class CsvDirectValue(val unit: CsvUnit) : CsvValueInterpretation

/**
 * The cell holds a mass, and the metric is that mass as a share of the row's
 * body weight. Scales export "Fat mass (kg)" while Health Connect stores a
 * percentage, so the weight column of the same row is needed.
 */
data class CsvMassShareOfWeight(
    /** Always a mass unit. */
    val unit: CsvUnit,
) : CsvValueInterpretation

/** Everything the importer needs to know about one metric. */
data class CsvMetricSpec(
    /** The Health Connect record class name. Part of the clientRecordId; must match Flutter. */
    val targetType: String,

    /** The record class, for the existing-id lookup. */
    val recordType: KClass<out Record>,

    /** The single Health Connect write permission this metric needs. */
    val writePermission: String,

    /** Offered in the UI in this order; the first is the default. */
    val interpretations: List<CsvValueInterpretation>,

    /** Bounds on the canonical value. A 900 kg weight is a mis-mapped column. */
    val plausibleMin: Double,
    val plausibleMax: Double,

    /** Whether the value covers a span. Reads the END_TIMESTAMP column; one minute without it. */
    val isInterval: Boolean = false,
) {
    val defaultInterpretation: CsvValueInterpretation get() = interpretations.first()
}

/** The metric catalog. Adding a metric is an entry here plus a label string. */
val CsvMetricCatalog: Map<CsvImportMetric, CsvMetricSpec> = mapOf(
    CsvImportMetric.WEIGHT to CsvMetricSpec(
        targetType = "WeightRecord",
        recordType = WeightRecord::class,
        writePermission = HealthPermission.getWritePermission(WeightRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.KILOGRAMS),
            CsvDirectValue(CsvUnit.POUNDS),
            CsvDirectValue(CsvUnit.STONES),
            CsvDirectValue(CsvUnit.GRAMS),
        ),
        plausibleMin = 2.0,
        plausibleMax = 650.0,
    ),
    CsvImportMetric.BODY_FAT to CsvMetricSpec(
        targetType = "BodyFatRecord",
        recordType = BodyFatRecord::class,
        writePermission = HealthPermission.getWritePermission(BodyFatRecord::class),
        // The mass interpretations make a Withings-style export usable as is.
        interpretations = listOf(
            CsvDirectValue(CsvUnit.PERCENT),
            CsvDirectValue(CsvUnit.FRACTION),
            CsvMassShareOfWeight(CsvUnit.KILOGRAMS),
            CsvMassShareOfWeight(CsvUnit.POUNDS),
            CsvMassShareOfWeight(CsvUnit.GRAMS),
        ),
        // Below 2% or above 75% means the wrong weight column was used.
        plausibleMin = 2.0,
        plausibleMax = 75.0,
    ),
    CsvImportMetric.LEAN_BODY_MASS to CsvMetricSpec(
        targetType = "LeanBodyMassRecord",
        recordType = LeanBodyMassRecord::class,
        writePermission = HealthPermission.getWritePermission(LeanBodyMassRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.KILOGRAMS),
            CsvDirectValue(CsvUnit.POUNDS),
            CsvDirectValue(CsvUnit.GRAMS),
        ),
        plausibleMin = 1.0,
        plausibleMax = 400.0,
    ),
    CsvImportMetric.BONE_MASS to CsvMetricSpec(
        targetType = "BoneMassRecord",
        recordType = BoneMassRecord::class,
        writePermission = HealthPermission.getWritePermission(BoneMassRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.KILOGRAMS),
            CsvDirectValue(CsvUnit.POUNDS),
            CsvDirectValue(CsvUnit.GRAMS),
        ),
        plausibleMin = 0.1,
        plausibleMax = 20.0,
    ),
    CsvImportMetric.BODY_WATER_MASS to CsvMetricSpec(
        targetType = "BodyWaterMassRecord",
        recordType = BodyWaterMassRecord::class,
        writePermission = HealthPermission.getWritePermission(BodyWaterMassRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.KILOGRAMS),
            CsvDirectValue(CsvUnit.POUNDS),
            CsvDirectValue(CsvUnit.GRAMS),
        ),
        plausibleMin = 0.5,
        plausibleMax = 400.0,
    ),
    CsvImportMetric.HEIGHT to CsvMetricSpec(
        targetType = "HeightRecord",
        recordType = HeightRecord::class,
        writePermission = HealthPermission.getWritePermission(HeightRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.CENTIMETERS),
            CsvDirectValue(CsvUnit.METERS),
            CsvDirectValue(CsvUnit.INCHES),
            CsvDirectValue(CsvUnit.FEET),
        ),
        // Canonical height is METRES.
        plausibleMin = 0.3,
        plausibleMax = 2.8,
    ),
    CsvImportMetric.BASAL_METABOLIC_RATE to CsvMetricSpec(
        targetType = "BasalMetabolicRateRecord",
        recordType = BasalMetabolicRateRecord::class,
        writePermission = HealthPermission.getWritePermission(BasalMetabolicRateRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.KILOCALORIES_PER_DAY),
            CsvDirectValue(CsvUnit.KILOJOULES_PER_DAY),
        ),
        plausibleMin = 200.0,
        plausibleMax = 12000.0,
    ),

    // Vitals. Ranges are "survivable human", not "clinically normal".
    CsvImportMetric.HEART_RATE to CsvMetricSpec(
        targetType = "HeartRateRecord",
        recordType = HeartRateRecord::class,
        writePermission = HealthPermission.getWritePermission(HeartRateRecord::class),
        interpretations = listOf(CsvDirectValue(CsvUnit.BEATS_PER_MINUTE)),
        plausibleMin = 1.0,
        plausibleMax = 300.0,
    ),
    CsvImportMetric.RESTING_HEART_RATE to CsvMetricSpec(
        targetType = "RestingHeartRateRecord",
        recordType = RestingHeartRateRecord::class,
        writePermission = HealthPermission.getWritePermission(RestingHeartRateRecord::class),
        interpretations = listOf(CsvDirectValue(CsvUnit.BEATS_PER_MINUTE)),
        plausibleMin = 1.0,
        plausibleMax = 300.0,
    ),
    CsvImportMetric.HEART_RATE_VARIABILITY to CsvMetricSpec(
        targetType = "HeartRateVariabilityRmssdRecord",
        recordType = HeartRateVariabilityRmssdRecord::class,
        writePermission = HealthPermission.getWritePermission(HeartRateVariabilityRmssdRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.MILLISECONDS),
            CsvDirectValue(CsvUnit.SECONDS),
        ),
        // Health Connect rejects RMSSD outside 1..200 ms.
        plausibleMin = 1.0,
        plausibleMax = 200.0,
    ),
    CsvImportMetric.OXYGEN_SATURATION to CsvMetricSpec(
        targetType = "OxygenSaturationRecord",
        recordType = OxygenSaturationRecord::class,
        writePermission = HealthPermission.getWritePermission(OxygenSaturationRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.PERCENT),
            CsvDirectValue(CsvUnit.FRACTION),
        ),
        plausibleMin = 50.0,
        plausibleMax = 100.0,
    ),
    CsvImportMetric.RESPIRATORY_RATE to CsvMetricSpec(
        targetType = "RespiratoryRateRecord",
        recordType = RespiratoryRateRecord::class,
        writePermission = HealthPermission.getWritePermission(RespiratoryRateRecord::class),
        interpretations = listOf(CsvDirectValue(CsvUnit.BREATHS_PER_MINUTE)),
        plausibleMin = 1.0,
        plausibleMax = 100.0,
    ),
    CsvImportMetric.BODY_TEMPERATURE to CsvMetricSpec(
        targetType = "BodyTemperatureRecord",
        recordType = BodyTemperatureRecord::class,
        writePermission = HealthPermission.getWritePermission(BodyTemperatureRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.CELSIUS),
            CsvDirectValue(CsvUnit.FAHRENHEIT),
        ),
        plausibleMin = 25.0,
        plausibleMax = 45.0,
    ),
    CsvImportMetric.BASAL_BODY_TEMPERATURE to CsvMetricSpec(
        targetType = "BasalBodyTemperatureRecord",
        recordType = BasalBodyTemperatureRecord::class,
        writePermission = HealthPermission.getWritePermission(BasalBodyTemperatureRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.CELSIUS),
            CsvDirectValue(CsvUnit.FAHRENHEIT),
        ),
        plausibleMin = 25.0,
        plausibleMax = 45.0,
    ),
    CsvImportMetric.BLOOD_GLUCOSE to CsvMetricSpec(
        targetType = "BloodGlucoseRecord",
        recordType = BloodGlucoseRecord::class,
        writePermission = HealthPermission.getWritePermission(BloodGlucoseRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.MILLIMOLES_PER_LITER),
            CsvDirectValue(CsvUnit.MILLIGRAMS_PER_DECILITER),
        ),
        // Canonical is mmol/L: roughly 18..900 mg/dL.
        plausibleMin = 1.0,
        plausibleMax = 50.0,
    ),
    CsvImportMetric.VO2_MAX to CsvMetricSpec(
        targetType = "Vo2MaxRecord",
        recordType = Vo2MaxRecord::class,
        writePermission = HealthPermission.getWritePermission(Vo2MaxRecord::class),
        interpretations = listOf(
            CsvDirectValue(CsvUnit.MILLILITERS_PER_KG_PER_MINUTE),
        ),
        plausibleMin = 5.0,
        plausibleMax = 100.0,
    ),

    // Intervals. A steps cell counts what happened between start and end.
    CsvImportMetric.STEPS to CsvMetricSpec(
        targetType = "StepsRecord",
        recordType = StepsRecord::class,
        writePermission = HealthPermission.getWritePermission(StepsRecord::class),
        interpretations = listOf(CsvDirectValue(CsvUnit.COUNT)),
        // Health Connect refuses a count below 1; 200,000 a day is past any ultramarathon.
        plausibleMin = 1.0,
        plausibleMax = 200000.0,
        isInterval = true,
    ),
)

/** Converts [value] from [unit] to the metric's canonical (metric) unit. */
fun convertCsvValueToCanonical(value: Double, unit: CsvUnit): Double = when (unit) {
    CsvUnit.KILOGRAMS -> value
    CsvUnit.POUNDS -> value / PoundsPerKilogram
    CsvUnit.STONES -> value * KilogramsPerStone
    CsvUnit.GRAMS -> value / 1000
    CsvUnit.PERCENT -> value
    CsvUnit.FRACTION -> value * 100
    CsvUnit.CENTIMETERS -> value / 100
    CsvUnit.METERS -> value
    CsvUnit.INCHES -> value * CentimetersPerInch / 100
    CsvUnit.FEET -> value * CentimetersPerFoot / 100
    CsvUnit.KILOCALORIES_PER_DAY -> value
    CsvUnit.KILOJOULES_PER_DAY -> value / KilojoulesPerKilocalorie
    CsvUnit.CELSIUS -> value
    CsvUnit.FAHRENHEIT -> (value - FahrenheitFreezingPoint) / FahrenheitPerCelsius
    CsvUnit.BEATS_PER_MINUTE -> value
    CsvUnit.MILLISECONDS -> value
    CsvUnit.SECONDS -> value * 1000
    CsvUnit.BREATHS_PER_MINUTE -> value
    CsvUnit.MILLIMOLES_PER_LITER -> value
    CsvUnit.MILLIGRAMS_PER_DECILITER -> value / MilligramsPerDeciliterPerMillimolePerLiter
    CsvUnit.MILLILITERS_PER_KG_PER_MINUTE -> value
    CsvUnit.COUNT -> value
}

/** Unit tokens recognised in a header, longest first so `kcal` beats `cal`. */
private val HeaderUnitTokens: List<Pair<String, CsvUnit>> = listOf(
    "kilograms" to CsvUnit.KILOGRAMS,
    "kilogram" to CsvUnit.KILOGRAMS,
    "pounds" to CsvUnit.POUNDS,
    "ml/kg/min" to CsvUnit.MILLILITERS_PER_KG_PER_MINUTE,
    "mmol/l" to CsvUnit.MILLIMOLES_PER_LITER,
    "mg/dl" to CsvUnit.MILLIGRAMS_PER_DECILITER,
    "breaths/min" to CsvUnit.BREATHS_PER_MINUTE,
    "brpm" to CsvUnit.BREATHS_PER_MINUTE,
    "kcal" to CsvUnit.KILOCALORIES_PER_DAY,
    "kj" to CsvUnit.KILOJOULES_PER_DAY,
    "bpm" to CsvUnit.BEATS_PER_MINUTE,
    "lbs" to CsvUnit.POUNDS,
    "lb" to CsvUnit.POUNDS,
    "st" to CsvUnit.STONES,
    "kg" to CsvUnit.KILOGRAMS,
    "cm" to CsvUnit.CENTIMETERS,
    "ms" to CsvUnit.MILLISECONDS,
    "°c" to CsvUnit.CELSIUS,
    "°f" to CsvUnit.FAHRENHEIT,
    "in" to CsvUnit.INCHES,
    "ft" to CsvUnit.FEET,
    "%" to CsvUnit.PERCENT,
    "c" to CsvUnit.CELSIUS,
    "f" to CsvUnit.FAHRENHEIT,
    "g" to CsvUnit.GRAMS,
    "m" to CsvUnit.METERS,
    "s" to CsvUnit.SECONDS,
)

private val HeaderUnitTailRegex = Regex("""\(([^()]*)\)\s*$""")

/**
 * The unit named in [header], or null. Only the parenthesised tail counts, so
 * "Weight in grams of food" is not read as grams.
 */
fun detectCsvUnitInHeader(header: String): CsvUnit? {
    val match = HeaderUnitTailRegex.find(header.trim()) ?: return null
    val inner = match.groupValues[1].trim().lowercase()
    if (inner.isEmpty()) return null
    for ((token, unit) in HeaderUnitTokens) {
        if (inner == token) return unit
    }
    return null
}

/** [spec]'s interpretation for [unit], preferring direct over derived. Null if none. */
fun interpretationForUnit(spec: CsvMetricSpec, unit: CsvUnit): CsvValueInterpretation? {
    spec.interpretations.firstOrNull { it is CsvDirectValue && it.unit == unit }?.let { return it }
    return spec.interpretations.firstOrNull { it is CsvMassShareOfWeight && it.unit == unit }
}
